keanu-project/src/main/java/io/improbable/keanu/algorithms/mcmc/HamiltonianSampler.java

```diff
@@ -0,0 +1,263 @@
+        step();
+        addSampleFromCache(samples, acceptOrReject());
+        logOfMasterPForEachSample.add(logOfMasterPBeforeLeapfrog);
+    }
+
+    @Override
+    public NetworkState sample() {
+        step();
+        return new SimpleNetworkState(acceptOrReject());
+    }
+
+    private Map<VertexId, ?> acceptOrReject() {
+        final double logOfMasterPAfterLeapfrog = bayesNet.getLogOfMasterP();
+
+        final double likelihoodOfLeapfrog = getLikelihoodOfLeapfrog(
+            logOfMasterPAfterLeapfrog,
+            logOfMasterPBeforeLeapfrog,
+            during.momentum,
+            before.momentum
+        );
+
+        if (shouldReject(likelihoodOfLeapfrog, random)) {
+
+            //Revert to position and gradient before leapfrog
+            Map<VertexId, DoubleTensor> tempSwap = during.position;
+            during.position = before.position;
+            before.position = tempSwap;
+
+            tempSwap = during.gradient;
+            during.gradient = before.gradient;
+            before.gradient = tempSwap;
+
+            return sampleBeforeLeapfrog;
+        } else {
+            logOfMasterPBeforeLeapfrog = logOfMasterPAfterLeapfrog;
+            return putSample(fromVertices);
+        }
+    }
+
+    private static void initializeMomentumForEachVertex(List<Vertex<DoubleTensor>> vertexes,
+                                                        Map<VertexId, DoubleTensor> momentums,
+                                                        KeanuRandom random) {
+        for (Vertex<DoubleTensor> currentVertex : vertexes) {
+            momentums.put(currentVertex.getId(), random.nextGaussian(currentVertex.getShape()));
+        }
+    }
+
+    /**
+     * function Leapfrog(T, r)
+     * Set `r = r + (eps/2)dTL(T)
+     * Set `T = T + r`
+     * Set `r = r` + (eps/2)dTL(`T)
+     * return `T, r`
+     *
+     * @param latentVertices
+     * @param during          current vertex position, gradient and momentum
+     * @param stepSize
+     * @param logProbGradient calculator of the logProb gradients
+     * @return the gradient at the updated position
+     */
+    private static Map<VertexId, DoubleTensor> leapfrog(final List<Vertex<DoubleTensor>> latentVertices,
+                                                        final VertexState during,
+                                                        final double stepSize,
+                                                        final LogProbGradientCalculator logProbGradient) {
+
+        final double halfTimeStep = stepSize / 2.0;
+
+        Map<VertexId, DoubleTensor> momentumsAtHalfTimeStep = new HashMap<>();
+
+        //Set `r = r + (eps/2)dTL(T)
+        for (Map.Entry<VertexId, DoubleTensor> vertexMomentum : during.momentum.entrySet()) {
+            final DoubleTensor updatedMomentum = during.gradient.get(vertexMomentum.getKey()).times(halfTimeStep).plusInPlace(vertexMomentum.getValue());
```

Comment:
I think we could probably make this much more readable by extracting out some variables...

Reply:
Actually this is true for all of the for loops here